### PR TITLE
DRILL-7030: Make format plugins fully pluggable

### PR DESCRIPTION
**Pull request by agozhiy**
- Bootstrap files for format plugins were introduced and added to the existing plugins in contrib.
- Formats from these files are being added dynamically to the corresponding storage plugins.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 1]
```diff
@@ -310,22 +311,18 @@ public void addPluginToPersistentStoreIfAbsent(String name, StoragePluginConfig
...
-        StoragePlugins plugins = lpPersistence.getMapper().readValue(pluginsData, StoragePlugins.class);
-        for (Entry<String, StoragePluginConfig> plugin : plugins) {
-          StoragePluginConfig oldPluginConfig = bootstrapPlugins.putIfAbsent(plugin.getKey(), plugin.getValue());
-          if (oldPluginConfig != null) {
-            logger.warn("Duplicate plugin instance '{}' defined in [{}, {}], ignoring the later one.",
-                plugin.getKey(), pluginURLMap.get(plugin.getKey()), url);
-          } else {
-            pluginURLMap.put(plugin.getKey(), url);
-          }
+      for (URL url : storageUrls) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: add log info, similar as for storage plugins

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 2]
```diff
@@ -334,6 +331,48 @@ private StoragePlugins loadBootstrapPlugins(LogicalPlanPersistence lpPersistence
     }
   }
 
+  private void loadStoragePlugins(URL url, StoragePlugins bootstrapPlugins, Map<String, URL> pluginURLMap, LogicalPlanPersistence lpPersistence) throws IOException {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Please add Javadoc.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 3]
```diff
@@ -334,6 +331,48 @@ private StoragePlugins loadBootstrapPlugins(LogicalPlanPersistence lpPersistence
...
+    plugins.forEach(formatPlugin -> {
+      String targetStoragePluginName = formatPlugin.getKey();
+      StoragePluginConfig storagePlugin = bootstrapPlugins.getConfig(targetStoragePluginName);
+      StoragePluginConfig formatPluginValue = formatPlugin.getValue();
+      if (storagePlugin == null) {
+        logger.warn("No storage plugins with the given name are registered: '{}'", targetStoragePluginName);
+      } else if (storagePlugin instanceof FileSystemConfig && formatPluginValue instanceof FileSystemConfig) {
+        FileSystemConfig targetPlugin = (FileSystemConfig) storagePlugin;
+        ((FileSystemConfig) formatPluginValue).getFormats().forEach((formatName, formatValue) -> {
+          if (targetPlugin.getFormats().containsKey(formatName)) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Use putIfAbsent as in `loadStoragePlugins`

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 4]
```diff
@@ -334,6 +331,48 @@ private StoragePlugins loadBootstrapPlugins(LogicalPlanPersistence lpPersistence
...
+            logger.warn("Duplicate format instance '{}' defined in [{}, {}], ignoring the later one.",
+                formatName, pluginURLMap.get(targetStoragePluginName), url);
+          } else {
+            targetPlugin.getFormats().put(formatName, formatValue);
+          }
+        });
+      } else {
+        logger.warn("Formats are only supported by File System plugin type: '{}'", targetStoragePluginName);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]:
```suggestion
        logger.warn("Formats are only supported by File System plugin type: [{}]", targetStoragePluginName);
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 5]
```diff
@@ -334,6 +331,48 @@ private StoragePlugins loadBootstrapPlugins(LogicalPlanPersistence lpPersistence
...
+  }
+
+  private void loadFormatPlugins(URL url, StoragePlugins bootstrapPlugins, Map<String, URL> pluginURLMap, LogicalPlanPersistence lpPersistence) throws IOException {
+    StoragePlugins plugins = getPluginsFromResource(url, lpPersistence);
+    plugins.forEach(formatPlugin -> {
+      String targetStoragePluginName = formatPlugin.getKey();
+      StoragePluginConfig storagePlugin = bootstrapPlugins.getConfig(targetStoragePluginName);
+      StoragePluginConfig formatPluginValue = formatPlugin.getValue();
+      if (storagePlugin == null) {
+        logger.warn("No storage plugins with the given name are registered: '{}'", targetStoragePluginName);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]:
```suggestion
        logger.warn("No storage plugins with the given name are registered: [{}]", targetStoragePluginName);
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 6]
```diff
@@ -334,6 +331,48 @@ private StoragePlugins loadBootstrapPlugins(LogicalPlanPersistence lpPersistence
     }
   }
 
+  private void loadStoragePlugins(URL url, StoragePlugins bootstrapPlugins, Map<String, URL> pluginURLMap, LogicalPlanPersistence lpPersistence) throws IOException {
+    StoragePlugins plugins = getPluginsFromResource(url, lpPersistence);
+    plugins.forEach(plugin -> {
+      StoragePluginConfig oldPluginConfig = bootstrapPlugins.putIfAbsent(plugin.getKey(), plugin.getValue());
+      if (oldPluginConfig != null) {
+        logger.warn("Duplicate plugin instance '{}' defined in [{}, {}], ignoring the later one.",
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]:
```suggestion
        logger.warn("Duplicate plugin instance [{}] defined in [{}, {}], ignoring the later one.",
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java [hunk 7]
```diff
@@ -334,6 +331,48 @@ private StoragePlugins loadBootstrapPlugins(LogicalPlanPersistence lpPersistence
...
+            plugin.getKey(), pluginURLMap.get(plugin.getKey()), url);
+      } else {
+        pluginURLMap.put(plugin.getKey(), url);
+      }
+    });
+  }
+
+  private void loadFormatPlugins(URL url, StoragePlugins bootstrapPlugins, Map<String, URL> pluginURLMap, LogicalPlanPersistence lpPersistence) throws IOException {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Please add Javadoc.

File: exec/java-exec/src/test/resources/bootstrap-format-plugins.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
+{
```
[the remark this answers is not in the document]
Comment (agozhiy) [reply exec/java-exec/src/test/resources/bootstrap-format-plugins.json]: They should be available for all tests.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSystemConfig.java [hunk 9]
```diff
@@ -48,7 +49,7 @@ public FileSystemConfig(@JsonProperty("connection") String connection,
     Map<String, WorkspaceConfig> caseInsensitiveWorkspaces = CaseInsensitiveMap.newHashMap();
     Optional.ofNullable(workspaces).ifPresent(caseInsensitiveWorkspaces::putAll);
     this.workspaces = caseInsensitiveWorkspaces;
-    this.formats = formats;
+    this.formats = formats != null? formats : new LinkedHashMap<>();
```
[the remark this answers is not in the document]
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSystemConfig.java]: Fixed.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java (hunk 1 above)
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java (hunk 2 above)
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java (hunk 3 above)
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java (hunk 4 above)
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java (hunk 5 above)
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java (hunk 6 above)
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java (hunk 7 above)
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java]: Done.

**Comment by agozhiy** [comment]
@arina-ielchiieva, I addressed your comments, could you please review?

File: exec/java-exec/src/test/resources/bootstrap-format-plugins.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (arina-ielchiieva) [reply exec/java-exec/src/test/resources/bootstrap-format-plugins.json]: Please try to use these files only for your unit tests so other tests do not get affected.

File: exec/java-exec/src/test/java/org/apache/drill/exec/store/TestFormatPluginLoading.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+  public TestFormatPluginLoading() throws Exception {
+    storagePluginRegistry = new StoragePluginRegistryImpl(mockDrillbitContext());
+    storagePluginRegistry.init();
+    plugins = storagePluginRegistry.getStore();
+  }
+
+  @BeforeClass
+  public static void setup() {
+    storagePlugins = new File(Resources.getResource("store/" + ExecConstants.BOOTSTRAP_STORAGE_PLUGINS_FILE).getFile());
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/store/TestFormatPluginLoading.java]: Can you please explain what you are doing here? I am sure what this renaming logic does.

**Comment by arina-ielchiieva** [comment]
+1, LGTM.

File: contrib/format-ltsv/src/main/resources/bootstrap-format-plugins.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (paul-rogers) [inline contrib/format-ltsv/src/main/resources/bootstrap-format-plugins.json]: Nice change. These new files add bootstrap config for several config plugins. Are there entries for these plugins in the "master" bootstrap file that should be removed?

Comment (agozhiy) [reply contrib/format-ltsv/src/main/resources/bootstrap-format-plugins.json]: No, they were not present.

**Pull request closed without merging**